libs/MeshKernel/include/MeshKernel/CasulliDeRefinement.hpp

```diff
@@ -0,0 +1,174 @@
+
+        /// @brief Compute coordinates of the new node.
+        Point ComputeNewNodeCoordinates(const Mesh2D& mesh,
+                                        const std::vector<int>& nodeTypes,
+                                        const UInt nodeId) const;
+
+        /// @brief Update the mesh members for the mesh description and connectivity.
+        [[nodiscard]] bool UpdateDirectlyConnectedElements(Mesh2D& mesh,
+                                                           const UInt elementId,
+                                                           const std::vector<UInt>& directlyConnected,
+                                                           const std::vector<std::array<int, 2>>& kne) const;
+
+        /// @brief Get the most significant node type for all nodes of the element.
+        int GetNodeCode(const Mesh2D& mesh,
+                        const std::vector<int>& nodeTypes,
+                        const UInt elementId) const;
+
+        /// @brief Add element id to the list of id's
+        ///
+        /// only added is it is not already on the list and the element is a quadrilateral
+        void AddElementToList(const Mesh& mesh, const std::vector<UInt>& frontList, std::vector<UInt>& frontListCopy, const UInt kNew) const;
+
+        /// @brief Redirect nodes of connected cells, deactivate polygons of degree smaller than three
+        void RedirectNodesOfConnectedElements(Mesh2D& mesh, const UInt elementId, const UInt nodeId, const std::vector<UInt>& indirectlyConnected) const;
+
+        /// @brief Removes nodes from the boundary that will not be part of the de-refined mesh.
+        [[nodiscard]] bool RemoveUnwantedBoundaryNodes(Mesh2D& mesh,
+                                                       const std::vector<int>& nodeTypes,
+                                                       const Polygons& polygon,
+                                                       const std::vector<UInt>& indirectlyConnected) const;
+
+        /// @brief Delete an element
+        [[nodiscard]] bool DeleteElement(Mesh2D& mesh,
+                                         std::vector<int>& nodeTypes,
+                                         const Polygons& polygon,
+                                         const UInt elementId,
+                                         const std::vector<UInt>& directlyConnected,
+                                         const std::vector<UInt>& indirectlyConnected,
+                                         const std::vector<std::array<int, 2>>& kne) const;
+
+        /// @brief Clean up the edge
+        ///
+        /// @returns Indicates if the cleanp-up was successful or not
+        [[nodiscard]] bool CleanUpEdge(Mesh2D& mesh, const UInt edgeId) const;
+
+        /// @brief Find the id of the shared node for two edges.
+        ///
+        /// If no node is shared then return null value
+        UInt FindCommonNode(const Mesh2D& mesh, const UInt edgeId1, const UInt edgeId2) const;
```

Comment:
Is this function already implemented elsewhere?
meshkernel::UInt Mesh::FindCommonNode(UInt firstEdgeIndex, UInt secondEdgeIndex) const